### ISSUE TITLE
rpm-package

### ISSUE THREAD

**Issue by Blaimi** [sezanzeb/input-remapper#83]
I'd love to see this project as an rpm-package for fedora.

**Comment by Blaimi** [comment]
I startet at https://github.com/Blaimi/key-mapper/tree/83_rpm-package :)

**Comment by sezanzeb** [comment]
Cool, but bear in mind that I can't test it because I don't use any fedora based distro

When would I need to worry about potentially breaking the rpm build with changes to key-mapper?

**Comment by Blaimi** [comment]
As long as you don't depend on newer versions of the python-libraries as they are in the repositories the should not be a problem. Since fedora has a faster release cycle than ubuntu (fedora has a ~6 month release cycle like ubuntu but without LTS-steps) I don't assume problems if the project fits the latest ubuntu LTS. Once the package is in the official repositories you cannot upload newer versions anyway for a certain release (i.e. if ubuntu 21.10 comes with key-mapper 1.0.0, you can upgrade the package to 1.1 not before ubuntu 22.04), but you can publish it in backports or a ppa.

I think if the fedora-package has problems with dependencies, you will have them on ubuntu as well. These problems should be solved when they pop up. That's another reason why I mentioned to split the packaging- from the source-codein #40. This way the package-stuff could even be done on the gitlab-instance of the debian-project (salsa.debian.org) where if can profit from their excellent CI-configurations.

**Comment by sezanzeb** [comment]
I can't find your comment on splitting the packaging from the source code

**Comment by Blaimi** [comment]
oh, sorry. it was this one: https://github.com/sezanzeb/key-mapper/pull/82#issuecomment-816592203. That's the difference between `3.0 (native)` and `3.0 (quilt)` in the file `debian/source/format`.

Of course you can. The benefits of salsa it that the package undergoes different tests automatically. like installations on different distributions and upgrades from old packages, etc.

But these discussions are not yes necessary. Ubuntu 21.04 is in beta, fedora 34 is in beta, debian 11 is in hard-freeze. I.e. the next deadline for upstream-packaging is at least 3 month ago. In the meantime we can try getting it into debian sid which migrates to testing which migrates to ubuntu, maybe set up a ppa and – if I get the fedora-rpm working – set up a copr (that's kind of ppa-equivalent in the fedora-world)

**Comment by sezanzeb** [comment]
I'm really glad there are people who know what all of that stuff is and how it works. Because I don't.

I didn't know there is a deadline, I'd love to meet it then.

Did I read this correctly that "quilt" means putting the deb building into a different repo on salsa.debian.org?

**Comment by Blaimi** [comment]
the format `quilt` got its name from the [software with the same name](https://savannah.nongnu.org/projects/quilt). It means that a maintainer can manipulate the projects source-code with patch-files saved inside the `debian`-directory. The reasons to do this is either because the software-developer and the package-maintainer don't know each other, the developer doesn't care about packages or, like in this case, to decouple/abstract the project-development from its packaging.

Splitting packaging from the software-project is very common and recommended. Using different repositories is optional and noone is forced to use salsa. I've seen already some projects using quilt inside the same repository with or without a debian-branch but this gets unclear and confusing very quick if you want to support more than one debian/ubuntu-version (e.g. [phpmyadmin](https://salsa.debian.org/phpmyadmin-team/phpmyadmin) currently has 9 branches on salsa – only for .deb-packaging).

You can also identify quilt-packages by their `-` in the version number, rpm even forces this.

A little example:
* native
  * You release version `1.2.3`
  * You find a bug in the packaging-code, e.g. in `debian/rules`
  * You are forced to release `1.2.4` for the fix – even for pip which is unaffected
* quilt
  * You release version `1.2.3`
  * You/the package-maintainer updates the packaging code for this release and releases `1.2.3-1`
  * You find a bug in the packaging-code
  * You fix the problem and release the fixed package as `1.2.3-2` – without touching the pip-release

This gets more and more complicated when supporting stable or different distributions or backports … that's the reason for version numbers like `1:10.3.22-1ubuntu1`. (first non-default epoch, upstream version 10.3.22, first debian-package-release, first ubuntu-additions-release) or `4.6.6-4+deb9u1` (upstream version 4.6.6, fourth package-release, first security-update for debian9).

native packaging is usually only used for packages which are highly bundled with the debian-ecosystem like `apt` or `devscripts` where it makes no sense to even think about porting them to a different system.

So, long story short:
No, It doesn't mean putting the code onto salsa. Using `quilt` is recommended but not necessary. If the packaging code uses quit I recommend to do this in a separate git-repository to abstract stuff and be less confusing. If this separate git-repo shall benefit from salsa's packaging-CI, at least a clone has to be hosted there.

**Comment by sezanzeb** [comment]
Thanks a lot

**Comment by sezanzeb** [comment]
Is it working for you as rpm package, do you have something to add in a PR? What's the state of this :)

**Comment by Blaimi** [comment]
I've not done much on that :/. My level of suffering is not very high since my primary device runs ubuntu and I'm usually working without mouse on my fedora-laptop. So: No, it's not working and no, you don't have to do something ;).

**Comment by andrewaylett** [comment]
I built an RPM package "by hand" using bdist:

```
touch inputremapper/commit_hash.py
python setup.py bdist_rpm
```

It doesn't out of the box know about its dependencies: you need to install them manually:

```
sudo dnf install python3-pydbus python3-evdev
sudo dnf install dist/input-remapper-*.noarch.rpm
```

Ensure the service is enabled:

```
sudo systemctl enable --now input-remapper.service
```

And you should find InputRemapper in your desktop environment's application launcher.

Once I'd worked out that the reason it wasn't "just working" was because setup.py was creating extra files that bdist didn't know about, it _just worked_.  Really nice job on the packaging, thank you.  And thank you for the software -- it's working perfectly.

**Comment by sunwire** [comment]
Repo https://copr.fedorainfracloud.org/coprs/sunwire/input-remapper/

**Comment by sezanzeb** [comment]
Thanks. Is copr like the AUR? I never used Fedora, I'm not familiar with any of this

I think it would be really cool if there was a script that builds an .rpm file in https://github.com/sezanzeb/input-remapper/tree/main/scripts

**Comment by sunwire** [comment]
> Thanks. Is copr like the AUR?

Yes

> I think it would be really cool if there was a script that builds an .rpm file

Where is no script, only spec file and source rpm (it contains all source code + spec file)
[Spec file](https://download.copr.fedorainfracloud.org/results/sunwire/input-remapper/fedora-35-x86_64/03358993-python-input-remapper/python-input-remapper.spec)
[Source RPM](https://download.copr.fedorainfracloud.org/results/sunwire/input-remapper/fedora-35-x86_64/03358993-python-input-remapper/python-input-remapper-1.4.0-3.git20220205.fc35.src.rpm)

**Comment by sezanzeb** [comment]
We could add the spec file into the repo (similar to how the deb files are stored in it https://github.com/sezanzeb/input-remapper/tree/main/DEBIAN), and the script would put everything into the .rpm archive and place it in `dist/input-remapper-1.4.0.rpm` or something.

Would make it easy to install local-changes of input-remapper on fedora. I also do that with the deb script when testing changes on ubuntu.

**Comment by sunwire** [comment]
> The script would put everything into the .rpm archive and place it in `dist/input-remapper-1.4.0.rpm` or something.

Well, it's not so easy. https://www.redhat.com/sysadmin/create-rpm-package
You have to (in short):
Setup dev environment
Install *.src.rpm
Install all dependencies
Edit the spec file to indicate a new commit
Download a new source code
Build a RPM package

**Comment by andrewaylett** [comment]
@sunwire, for Python packages like this one it's easy enough -- per my comment above, the script would look something like this:

```
touch inputremapper/commit_hash.py
python setup.py bdist_rpm
```

It's not going to be complete enough to submit to a distribution for inclusion, but it's good enough for a local install.

**Comment by sunwire** [comment]
> It's not going to be complete enough to submit to a distribution for inclusion, but it's good enough for a local install.

Generally I agree, but for an ordinary user building the RPM package can be a challenge.
The problem may also be missing dependencies that prevent the program from starting.
The advantage of RPM over PIP is that the system knows what dependencies are needed.
DNF or YUM can install them but it must know about them first. RPM created with bdist_rpm doesn't provide any dependencies.

I think that it is possible to create a spec file with the appropriate Requires and BuildRequires tags by creating a special setup.cfg file. See: https://docs.python.org/3/distutils/builtdist.html#creating-rpm-packages and https://docs.python.org/3/distutils/configfile.html#setup-config
But I'm not an expert on this matter.
Repo (e.g copr) imho has many advantages over manual building and installing the rpm package, especially for an ordinary user.

**Comment by sezanzeb** [comment]
> Repo https://copr.fedorainfracloud.org/coprs/sunwire/input-remapper/

hey @sunwire, input-remapper requires a newer python3-evdev version, how does that work for fedora?

**Comment by sunwire** [comment]
> hey @sunwire, input-remapper requires a newer python3-evdev version, how does that work for fedora?

It doesn't :( python3-evdev is too old bug report is filled https://bugzilla.redhat.com/show_bug.cgi?id=1985644 but the maintainer doesn't respond.

**Comment by sezanzeb** [comment]
@gvalkov doesn't seem to actively maintain it anymore. This includes the repo: https://github.com/gvalkov/python-evdev/pulls

**Comment by sezanzeb** [comment]
Considering the widespread usage of python-evdev, someone should be added as a maintainer there.

And someone will likely have to take over the python3-evdev Fedora package.

**Comment by sunwire** [comment]
> But I have no Idea how Fedora stuff is organized, or where to report that a package is not maintained anymore.

https://docs.fedoraproject.org/en-US/fesco/Policy_for_nonresponsive_package_maintainers/

> I unsuccessfully tried to reach out to him in the past already.

I am also waiting for a response from him.

**Comment by sezanzeb** [comment]
Thanks. Is it possible to turn this into an official Fedora package at some point? https://docs.fedoraproject.org/en-US/package-maintainers/Joining_the_Package_Maintainers/

**Comment by sezanzeb** [comment]
I was added as maintainer for the python-evdev github repo now and intend to take care of future pull requests there. Hopefully he will update the Fedora package soon as well or find someone to help out on that front :)

**Comment by sunwire** [comment]
Georgi contacted me and I have good news, the new version of python-evdev (1.5.0) will be available in Fedora soon.
https://bodhi.fedoraproject.org/updates/FEDORA-2022-038bec69b5

**Comment by Blaimi** [comment]
closing as duplicate in favour of #83 because it has more progress.
